### PR TITLE
[Validator] deprecate the "allowEmptyString" option

### LINKED ISSUE #36812: [Validator] LengthValidator: allowEmptyString option set to false allows empty strings
**Symfony version(s) affected**: 4.4.7

**Description**  
**allowEmptyString** option of Symfony\Component\Validator\Constraints\LengthValidator allows empty string

**How to reproduce**  
1. Create a form extending Symfony\Component\Form\AbstractType
2. Inside buildForm() method, add a field

```
$builder
    ...
    ->add('title', TextType::class, [
        'constraints' => [
            new Length([
                'allowEmptyString' => false,
                'max' => 255,
            ])
        ]
    ])
    ...
```

3. According to documentation
https://symfony.com/doc/4.4/reference/constraints/Length.html#allowemptystring
> Set it to false to consider empty strings not valid.

4. Create a controller, submit the form described above with data
```
['title' => '']
```

**Expectation:**
Form is invalid and has a proper error.

**Reality:**
The form is valid.

**The Reason**  
Currently, the code in **LengthValidator** only has the following lines mentioning **allowEmptyString**:

```
if (null !== $constraint->min && null === $constraint->allowEmptyString) {
    @trigger_error(sprintf('Using the "%s" constraint with the "min" option without setting the "allowEmptyString" one is deprecated and defaults to true. In 5.0, it will become optional and default to false.', Length::class), E_USER_DEPRECATED);
}

if (null === $value || ('' === $value && ($constraint->allowEmptyString ?? true))) {
    return;
}
```

The first block only triggers error if 'min' option is provided. In my example above 'min' is not provided.

The second block inner part (return) is not reached as well, because
```null === $value``` evaluates to ```false```
```('' === $value && ($constraint->allowEmptyString ?? true)))``` also evaluates to ```false```:
1) `'' === $value` evaluates to `true`
2) `$constraint->allowEmptyString ?? true` evaluates to `false`
3) `true && false` evaluates to `false`

Further down the code, **allowEmptyString** is never checked. Only **charset**, **min** and **max** validations are checked.

**Possible Solution** 
Write code which validates the the given string is not empty if **allowEmptyString** is false.
**Also, write tests.**

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | Fix #36812
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by derrabus** [comment]
Can we deprecate the `allowEmptyString` option instead? If someone wanted to disallow empty strings, they could use the `NotBlank` validator instead, or am I missing something?

**Comment by xabbuh** [comment]
Good idea, this was actually suggested when the option was introduced in 4.4 (see #31528):

> In my opinion, the flag could even be removed later.

File: src/Symfony/Component/Validator/Tests/Constraints/LengthTest.php [hunk 1]
```diff
@@ -39,4 +42,23 @@ public function testInvalidNormalizerObjectThrowsException()
         $this->expectExceptionMessage('The "normalizer" option must be a valid callable ("stdClass" given).');
         new Length(['min' => 0, 'max' => 10, 'normalizer' => new \stdClass()]);
     }
+
+    /**
+     * @group legacy
+     * @dataProvider allowEmptyStringOptionData
+     */
+    public function testDoNotAllowEmptyStringWithoutMinLength(bool $value)
```
Comment (derrabus) [inline src/Symfony/Component/Validator/Tests/Constraints/LengthTest.php:50]: The name of the test should be adjusted.

**Review by derrabus — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by xosofox** [comment]
Just wondering how we can achieve the "old behaviour" now when combining min, max and allowEmptyString=true

It used to be "ok" to have have either an empty string OR apply min-max-constraints - this is gone now. Or am I missing something?

**Comment by xabbuh** [comment]
Precisely for such use cases you can combine the `Length` constraint with other constraints using `AtLeastOne` (see the example in the upgrade guide): https://symfony.com/doc/current/reference/constraints/AtLeastOneOf.html

**Pull request merged**